app/assets/src/api/index.js

```diff
@@ -251,7 +251,56 @@ const createProject = params =>
 const logAnalyticsEvent = (eventName, eventData = {}) => {
   // Wrapper around Segment analytics so we can add things later
   // eventData should have keys in snake_case for the database
-  if (window.analytics) window.analytics.track(eventName, eventData);
+  if (window.analytics) {
+    // Include high value user groups in event properties to avoid JOINs downstream.
+    if (window.analytics.user) {
+      const traits = window.analytics.user().traits();
+      eventData = {
+        // see traits_for_segment
+        admin: traits.admin,
+        biohub_user: traits.biohub_user,
+        czi_user: traits.biohub_user,
+        demo_user: traits.demo_user,
+        has_samples: traits.has_samples,
+        // caller can override above traits if they know what they are doing
+        ...eventData
+      };
+    }
+    window.analytics.track(eventName, eventData);
+  }
+};
+
+/**
+ * For wrapping event handlers in React. The first part of the name should be
+ * the React component. The middle part should be user-friendly UI name. The
+ * last part should be the action taken.
+ *
+ * <ReactComponent>_<friendly_name_of_ui_element>_<past_tense_action>
+ *
+ * In this way, we can easily discern the
+ * context and meaning of an analytics event in a report and locate it in the
+ * codebase.
+ *
+ * For example:
+ *
+ *    withAnalytics(
+ *      this.renderMoreReads,
+ *      "AccessionViz_more_reads_link_clicked",
+ *      { reads: this.state.reads.length, allReads: this.allReads.length }
+ *    )
+ *
+ * React events should have have a single callsite, so there is no need to put
+ * them in ANALYTICS_EVENT_NAMES.
+ *
+ * The eventData should be used for extra data that would be useful for an
+ * analysis specific to the event.
+ **/
+const withAnalytics = (handleEvent, eventName, eventData = {}) => {
+  return (...args) => {
+    const ret = handleEvent(...args);
+    logAnalyticsEvent(eventName, eventData);
```

Comment:
Shouldnt this be an async call? Or `withAnalytics` and async function?

Reply:
The segment API is synchronous and returns immediately. I believe events are put into a queue for dispatch by http request.

Reply:
I would still make it asynchronous (1) since it is easy to do to, (2) avoids any dependencies on segment API which we do not control, (3) we do not use the return value of the call.
Given your answer, it is not as critical... 

We should probably have a try catch block around handleEvent to still log the interaction. This to avoid logging before the handling...

Reply:
Maybe I'm not getting you completely, but I don't see how it's easy to make `withAnalytics` async. Wouldn't that require `await` before every call?

Reply:
No, you can call an async function and continue without waiting for it. 
You only need `await` if you are interested in the result.

Reply:
I'm going to merge this PR now so I can make some progress and I trust we can discuss the async question later.